Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,271 @@
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Score = score;
+        user.ScoreSubmitted = true;
+
+        SetUser(battleId, userAddress, user);
+
+        var ScoreSubmittedCount = GetScoreSubmittedCount(battleId);
+        ScoreSubmittedCount += 1;
+        SetScoreSubmittedCount(battleId, ScoreSubmittedCount);
+
+        var highestScorer = GetHighestScorer(battleId);
+
+        if (score > highestScorer.Score)
+            SetHighestScorer(battleId, new BattleHighestScorer { Scorers = new Address[] { userAddress }, Score = score });
+        else if (score == highestScorer.Score)
+        {
+            var scorers = highestScorer.Scorers;
+            Array.Resize(ref scorers, scorers.Length + 1);
+            scorers[scorers.Length - 1] = userAddress;
+            SetHighestScorer(battleId, new BattleHighestScorer { Scorers = scorers, Score = score });
+        }
+
+        if (ScoreSubmittedCount == MaxUsers)
+        {
+            highestScorer = GetHighestScorer(battleId);
+            if (highestScorer.Scorers.Length > 1)
+                CancelBattle(battle);
+            else
+                ProcessWinner(battle, highestScorer.Scorers[0]);
+        }
+
+        Log(new BattleEventLog { Event = "End", BattleId = battleId, Address = Message.Sender });
+    }
+    /// <summary>
+    /// Get winner address
+    /// </summary>
+    public Address GetWinner(ulong battleId)
+    {
+        var battle = GetBattle(battleId);
+        return battle.Winner;
+    }
+    /// <summary>
+    /// Process winner when all user scores are submitted
+    /// </summary>
+    private void ProcessWinner(BattleMain battle, Address winnerAddress)
+    {
+        battle.Winner = winnerAddress;
+        SetBattle(battle.BattleId, battle);
+        ProcessPrize(battle);
+    }
+    /// <summary>
+    /// Send 3/4 amount to winner and 1/4 amount to battle owner
+    /// </summary>
+    private void ProcessPrize(BattleMain battle)
+    {
+        var prize = battle.Fee * (MaxUsers - 1);
+        Transfer(battle.Winner, prize);
+        Transfer(Owner, battle.Fee);
+    }
+    /// <summary>
+    /// Cancel battle and refund the fee amount
+    /// </summary>
+    private void CancelBattle(BattleMain battle)
+    {
+        battle.IsCancelled = true;
+        SetBattle(battle.BattleId, battle);
+
+        Transfer(battle.Users[0], battle.Fee);
+        Transfer(battle.Users[1], battle.Fee);
+        Transfer(battle.Users[2], battle.Fee);
+        Transfer(battle.Users[3], battle.Fee);
+    }
+    private void SetBattle(ulong battleId, BattleMain battle)
+    {
+        State.SetStruct($"battle:{battleId}", battle);
+    }
+    public BattleMain GetBattle(ulong battleId)
+    {
+        return State.GetStruct<BattleMain>($"battle:{battleId}");
+    }
+    private void SetUser(ulong battleId, Address address, BattleUser user)
+    {
+        State.SetStruct($"user:{battleId}-{address}", user);
```

Comment:
Always use `:` for multiple key seperator and do this for others. 
```suggestion
        State.SetStruct($"user:{battleId}:{address}", user);
```

Reply:
Also these are getters and setters method over State so please move them to top of constructor.